third_party/move/move-vm/runtime/src/interpreter.rs

```diff
@@ -529,8 +538,89 @@ impl InterpreterImpl {
                         loader,
                         function,
                         frame_cache,
+                        ClosureMask::empty(),
+                        vec![],
                     )?;
                 },
+                ExitCode::CallClosure(_sig_idx) => {
+                    // TODO(#15664): runtime check whether closure value on stack is same as sig_idx
+                    let (fun, captured) = self
+                        .operand_stack
+                        .pop_as::<Closure>()
+                        .map_err(|e| set_err_info!(current_frame, e))?
+                        .unpack();
+                    let lazy_function = LazyLoadedFunction::expect_this_impl(fun.as_ref())
+                        .map_err(|e| set_err_info!(current_frame, e))?;
+                    let mask = lazy_function.closure_mask();
+                    let function = lazy_function
+                        .with_resolved_function(module_storage, |f| Ok(f.clone()))
+                        .map_err(|e| set_err_info!(current_frame, e))?;
+
+                    // In difference to regular calls, we skip visibility check.
+                    // It is possible to call a private function of another module via
+                    // a closure.
+
+                    // Charge gas for module loading. If this function has no associated
+                    // module, it stems from a script.
+                    // TODO(#15664): currently we need the module id for gas charging of calls,
+                    //   so we can't proceed here without one.
+                    let module_id = function.module_id().ok_or_else(|| {
+                        let err = PartialVMError::new_invariant_violation(format!(
+                            "module id required to charge gas for function `{}`",
+                            lazy_function.to_stable_string()
+                        ));
+                        set_err_info!(current_frame, err)
+                    })?;
+                    let arena_id = traversal_context
+                        .referenced_module_ids
+                        .alloc(module_id.clone());
+                    loader.check_dependencies_and_charge_gas(
```

Comment:
Should we charge the gas prior to function resolution?

Reply:
+1, yes, we should be able to get module id from both resolved/unresolved function, and as commented on the PR below, we can just charge gas inside `with_resolved_function` to ensure gas metering always runs BEFORE we load anything.

Reply:
Done so. Also now metering here type arguments of the function, as asked for by @georgemitenkov on the underlying PR.